test/Gen.hs

```diff
@@ -0,0 +1,176 @@
+{-# LANGUAGE DataKinds, DeriveFunctor, DeriveGeneric, FlexibleInstances, FunctionalDependencies, GADTs, GeneralizedNewtypeDeriving, LambdaCase, PatternSynonyms, PolyKinds, RankNTypes, ScopedTypeVariables, StandaloneDeriving, TypeApplications, TypeOperators, UndecidableInstances, ViewPatterns #-}
+module Gen
+( module Control.Carrier.Pure
+  -- * Polymorphic generation & instantiation
+, m
+, genT
+, a
+, b
+, e
+, r
+, s
+, w
+, T(..)
+, A
+, B
+, E
+, R
+, S
+, W
+  -- * Generation
+, Rec(..)
+, forall
+  -- * Showing generated values
+, With(getWith)
+, showing
+, showingFn
+, atom
+, liftWith
+, liftWith2
+  -- * Pattern synonyms
+, pattern With
+, pattern Fn
+, pattern FnWith
+  -- * Re-exports
+, Gen
+, (===)
+, (/==)
+, choice
+, subterm
+, subtermM
+, subterm2
+, Arg
+, Vary
+, fn
+, apply
+) where
+
+import Control.Carrier.Pure
+import Data.Function (on)
+import Data.Functor.Classes (showsUnaryWith)
+import Data.Proxy
+import GHC.Stack
+import GHC.TypeLits
+import Hedgehog
+import Hedgehog.Function hiding (R, S)
+import Hedgehog.Gen
+import Hedgehog.Range
+
+-- | A generator for computations, given a higher-order generator for effectful operations, & a generator for results.
+m
+  :: forall m a
+  .  (Applicative m, Show a)
+  => (forall a . Show a => (forall a . Show a => Gen a -> Gen (With (m a))) -> Gen a -> Gen (With (m a))) -- ^ A higher-order generator producing operations using any effects in @m@.
+  -> Gen a                                                                                      -- ^ A generator for results.
+  -> Gen (With (m a))                                                                           -- ^ A generator producing computations, wrapped in 'With' for convenience.
+m with = go where
+  go :: forall a . Show a => Gen a -> Gen (With (m a))
+  go a = recursive choice [(liftWith "pure" pure . showing) <$> a] [with go a]
+
+
+genT :: Gen (T a)
+genT = T <$> integral (linear 0 100)
+
+newtype T a = T { unT :: Integer }
+  deriving (Enum, Eq, Generic, Num, Ord, Real, Vary)
+
+instance Arg (T a)
+
+instance KnownSymbol s => Show (T s) where
+  showsPrec d = showsUnaryWith showsPrec (symbolVal (Proxy @s)) d . unT
+
+a :: Gen A
+a = genT
+
+type A = T "A"
+
+b :: Gen B
+b = genT
+
+type B = T "B"
+
+e :: Gen E
+e = genT
+
+type E = T "E"
+
+r :: Gen R
+r = genT
+
+type R = T "R"
+
+s :: Gen S
+s = genT
+
+type S = T "S"
+
+w :: Gen W
+w = list (linear 0 100) genT
+
+type W = [T "W"]
+
+
+infixr 5 :.
+
+data Rec as where
+  Nil :: Rec '[]
+  (:.) :: a -> Rec as -> Rec (a ': as)
+
+forall :: (Forall g f, HasCallStack) => g -> f -> Hedgehog.Property
+forall g f = withFrozenCallStack $ Hedgehog.property (forall' g f)
+
+class Forall g f | g -> f, f -> g where
+  forall' :: HasCallStack => g -> f -> PropertyT IO ()
```

Comment:
I missed the convenience QuickCheck offered of specifying a property as a function from the generated values to the (in)equality to be tested. `forall` takes a heterogeneous list (`Rec`) of generators, generates a value from each in turn, and calls its second argument with these values. This is far more succinct than generating values using `do`-notation, while still allowing the property to label and classify values.